hypothesis-python/src/hypothesis/internal/scrutineer.py

```diff
@@ -38,11 +39,45 @@ def trace(self, frame, event, arg):
         if event == "call":
             return self.trace
         elif event == "line":
-            fname = frame.f_code.co_filename
-            if should_trace_file(fname):
-                current_location = (fname, frame.f_lineno)
-                self.branches.add((self._previous_location, current_location))
-                self._previous_location = current_location
+            self.trace_line(frame.f_code, frame.f_lineno)
+
+    def trace_line(self, code: types.CodeType, line_number: int):
+        fname = code.co_filename
+        if should_trace_file(fname):
+            current_location = (fname, line_number)
+            self.branches.add((self._previous_location, current_location))
+            self._previous_location = current_location
+
+
+# where possible, we'll use 3.12's new sys.monitoring module for low-overhead
+# coverage instrumentation. Otherwise, we'll default to sys.settrace.
+# This can be simplified once we drop 3.11.
+# tool_id = 2 is designated for coverage.
+MONITORING_TOOL_ID = 2
+if sys.version_info[:2] >= (3, 12):
+    MONITORING_EVENTS = {sys.monitoring.events.LINE: "trace_line"}
+
+
+def settracer(tracer: Tracer):
```

Comment:
This is effectively formalizing `Tracer` as an abstraction over `sys.settrace` and `sys.monitoring`. An alternative is to make this even *more* formal, with an abstract class that fully implements the functionality of both, and is inherited by `Tracer`. This could allow hypofuzz to reuse said abstract class (https://github.com/Zac-HD/hypofuzz/issues/9), depending on the kind of reuse you were thinking about.

If you think this is a good direction, I'd be open to implementing it.

Reply:
See above for the context manger note - I think having two implementations we can use in the same way would be useful, but since I don't ever expect for there to be a third it seems overkill to use an abstract class.

Reply:
Thank you